ports/esp32s2/common-hal/socketpool/Socket.c

```diff
@@ -96,51 +166,60 @@ mp_uint_t common_hal_socketpool_socket_send(socketpool_socket_obj_t* self, const
 
 mp_uint_t common_hal_socketpool_socket_recv_into(socketpool_socket_obj_t* self, const uint8_t* buf, mp_uint_t len) {
     size_t received = 0;
-    int status = 0;
-    uint64_t start_ticks = supervisor_ticks_ms64();
-    int sockfd;
-    esp_err_t err = esp_tls_get_conn_sockfd(self->tcp, &sockfd);
-    if (err != ESP_OK) {
-        mp_raise_OSError(MP_EBADF);
-    }
-    while (received == 0 &&
-           status >= 0 &&
-           (self->timeout_ms == 0 || supervisor_ticks_ms64() - start_ticks <= self->timeout_ms) &&
-           !mp_hal_is_interrupted()) {
-        RUN_BACKGROUND_TASKS;
-        size_t available = esp_tls_get_bytes_avail(self->tcp);
-        if (available == 0) {
-            // This reads the raw socket buffer and is used for non-TLS connections
-            // and between encrypted TLS blocks.
-            status = lwip_ioctl(sockfd, FIONREAD, &available);
-        }
-        size_t remaining = len - received;
-        if (available > remaining) {
-            available = remaining;
+
+    if (self->num != -1) {
+        // LWIP Socket
+        received = lwip_recv(self->num, (void*) buf, len - 1, 0);
```

Comment:
I think there are a couple approaches I could take to this:
1) There aren't LWIP equivalents of the TLS functions you use, so I could attempt to create a non-blocking version of this call by recreating the contents of lwip_recvfrom, lwip_recv_tcp, and the netconn functions they contain. I'm not sure how practical this would be.
2) Use getsockopt to set a timeout for this function, start a new thread in FreeRTOS (I don't know if we have any existing examples of this in Circuitpython), and wait for the timeout. Searching the web doesn't show any options for actually terminating a call in process before it times out, though.
3) use getsockopt to set a timeout and accept that Circuitpython will hang until it is complete. This is the most easily implemented but least flexible.

This is definitely an area where I'd like to ask someone more experienced with socket work for their opinion.

Reply:
I believe you can set the socket to non-blocking in lwip. You shouldn't need to copy code out of lwip. You'll just need to figure out what recv returns when nothing is available.

Reply:
I've spent a couple days on this, along with the error issue, and I'm still quite stuck. Let's try and touch base when you have time.